File: 2023-04(백준, 프로그래머스)/표 편집.py
class node:
    def __init__(self,left = None,right = None):
        self.remove = False
        self.left = left
        self.right = right
def solution(n, k, cmds):
    link = [node(i-1,i+1) for i in range(n)]
    link[0].left = None
    link[-1].right = None
    now = k
    check = []

    for cmd in cmds:
        if cmd[0] == 'U':
            c,num = cmd.split()
            for _ in range(int(num)):
                now = link[now].left
        elif cmd[0] == 'D':
            c,num = cmd.split()
            for _ in range(int(num)):
                now = link[now].right
        elif cmd[0] == 'C':
            check.append(now)
            link[now].remove = True
            l,r = link[now].left, link[now].right

            if l != None:
                link[l].right = r
            if r:
                link[r].left = l
                now = r
            else:
                now = l
        else:
            p = check.pop()
            link[p].remove = False
            l,r = link[p].left, link[p].right
            if l:
                link[l].right = p
            if r:
                link[r].left = p
                
    result = ['X' if i.remove else 'O' for i in link]
    return ''.join(result)

File: JS 코테 연습/표 편집.js
function solution(n, k, cmd) {
    let cursor = k
    let link = [...Array(n).keys()].map(i => [i-1, i+1])
    let del_list = []
    
    for (let c of cmd){
        c = c.split(' ')
        switch (c[0]){
            case 'U':
                for (let i =0; i < c[1]; i++){
                    cursor = link[cursor][0]
                }
                break
            case 'D':
                for (let i =0; i < c[1]; i++){
                    cursor = link[cursor][1]
                }
                break
            case 'C':
                let t = link[cursor]
                del_list.push([cursor, t])
                if (t[0] !== -1) {
                    link[t[0]][1] = t[1]
                }
                if (t[1] !== n) {
                    link[t[1]][0] = t[0]
                }
                cursor = t[1] === n ? t[0] : t[1]
                break
            case 'Z':
                let [cur, lr] = del_list.pop()
                if (lr[0] !== -1) {
                    link[lr[0]][1] = cur
                }
                if (lr[1] !== n) {
                    link[lr[1]][0] = cur
                }
                break
        }
    }

    let result = [...Array(n).keys()].map(_ => 'O')
    for (let i of del_list){
        result[i[0]] = 'X'
    }
    return result.join('')
}